src/probability/kernel/cond_cdf.lean

```diff
@@ -0,0 +1,948 @@
+    { exact ennreal.zero_ne_top, }, },
+  { refine (tendsto_congr' _).mp tendsto_const_nhds,
+    rw [eventually_eq, eventually_at_bot],
+    refine ⟨-1, λ q hq, (if_pos (hq.trans_lt _)).symm⟩,
+    linarith, },
+end
+
+lemma tendsto_cond_cdf_rat_at_top (ρ : measure (α × ℝ)) (a : α) :
+  tendsto (cond_cdf_rat ρ a) at_top (𝓝 1) :=
+begin
+  unfold cond_cdf_rat,
+  split_ifs with h,
+  { have h' := has_cond_cdf_of_mem_cond_cdf_set h,
+    rw [← ennreal.one_to_real, ennreal.tendsto_to_real_iff],
+    { exact h'.tendsto_at_top_one, },
+    { exact λ r, ((h'.le_one r).trans_lt ennreal.one_lt_top).ne, },
+    { exact ennreal.one_ne_top, }, },
+  { refine (tendsto_congr' _).mp tendsto_const_nhds,
+    rw [eventually_eq, eventually_at_top],
+    exact ⟨0, λ q hq, (if_neg (not_lt.mpr hq)).symm⟩, },
+end
+
+lemma cond_cdf_rat_ae_eq (ρ : measure (α × ℝ)) [is_finite_measure ρ] (r : ℚ) :
+  (λ a, cond_cdf_rat ρ a r) =ᵐ[ρ.fst] λ a, (pre_cdf ρ r a).to_real :=
+by filter_upwards [mem_cond_cdf_set_ae ρ] with a ha using cond_cdf_rat_of_mem ρ a ha r
+
+lemma of_real_cond_cdf_rat_ae_eq (ρ : measure (α × ℝ)) [is_finite_measure ρ] (r : ℚ) :
+  (λ a, ennreal.of_real (cond_cdf_rat ρ a r)) =ᵐ[ρ.fst] pre_cdf ρ r :=
+begin
+  filter_upwards [cond_cdf_rat_ae_eq ρ r, pre_cdf_le_one ρ] with a ha ha_le_one,
+  rw [ha, ennreal.of_real_to_real],
+  exact ((ha_le_one r).trans_lt ennreal.one_lt_top).ne,
+end
+
+lemma inf_gt_cond_cdf_rat (ρ : measure (α × ℝ)) (a : α) (t : ℚ) :
+  (⨅ r : Ioi t, cond_cdf_rat ρ a r) = cond_cdf_rat ρ a t :=
+begin
+  by_cases ha : a ∈ cond_cdf_set ρ,
+  { simp_rw cond_cdf_rat_of_mem ρ a ha,
+    have ha' := has_cond_cdf_of_mem_cond_cdf_set ha,
+    rw ← ennreal.to_real_infi,
+    { suffices : (⨅ (i : ↥(Ioi t)), pre_cdf ρ ↑i a) = pre_cdf ρ t a, by rw this,
+      rw ← ha'.infi_rat_gt_eq, },
+    { exact λ r, ((ha'.le_one r).trans_lt ennreal.one_lt_top).ne, }, },
+  { simp_rw cond_cdf_rat_of_not_mem ρ a ha,
+    have h_bdd : bdd_below (range (λ (r : ↥(Ioi t)), ite ((r : ℚ) < 0) (0 : ℝ) 1)),
+    { refine ⟨0, λ x hx, _⟩,
+      obtain ⟨y, rfl⟩ := mem_range.mpr hx,
+      dsimp only,
+      split_ifs,
+      exacts [le_rfl, zero_le_one], },
+    split_ifs with h h,
+    { refine le_antisymm _ (le_cinfi (λ x, _)),
+      { obtain ⟨q, htq, hq_neg⟩ : ∃ q, t < q ∧ q < 0,
+        { refine ⟨t/2, _, _⟩,
+          { linarith, },
+          { linarith, }, },
+        refine (cinfi_le h_bdd ⟨q, htq⟩).trans _,
+        rw if_pos,
+        rwa subtype.coe_mk, },
+      { split_ifs,
+        exacts [le_rfl, zero_le_one], }, },
+    { refine le_antisymm _ _,
+      { refine (cinfi_le h_bdd ⟨t+1, lt_add_one t⟩).trans _,
+        split_ifs,
+        exacts [zero_le_one, le_rfl], },
+      { refine le_cinfi (λ x, _),
+        rw if_neg,
+        rw not_lt at h ⊢,
+        exact h.trans (mem_Ioi.mp x.prop).le, }, }, },
+end
+
+/-- Conditional cdf of the measure given the value on `α`, as a plain function. This is an auxiliary
```

Comment:
Can you make it irreducible? It is a complicated (and noncanonical) definition, so no-one should ever need to open it once the API is in place, and it can make Lean's job easier.